garage/tf/models/base.py

```diff
@@ -0,0 +1,313 @@
+"""Base model classes."""
+import abc
+import warnings
+
+import tensorflow as tf
+
+
+class BaseModel(abc.ABC):
+    """Interface-only abstract class for models."""
+
+    def build(self, *inputs):
+        """
+        Output of model with the given input placeholder(s).
+
+        This function is implemented by subclasses to create their computation
+        graphs, which will be managed by Model. Generally, subclasses should
+        implement `build()` directly.
+        """
+        pass
+
+    @property
+    def name(self):
+        """Name for this Model."""
+        pass
+
+    @property
+    def input(self):
+        """Tensor input of the Model."""
+        pass
+
+    @property
+    def output(self):
+        """Tensor output of the Model."""
+        pass
+
+    @property
+    def inputs(self):
+        """Tensor inputs of the Model."""
+        pass
+
+    @property
+    def outputs(self):
+        """Tensor outputs of the Model."""
+        pass
+
+    @property
+    def parameters(self):
+        """Parameters of the Model."""
+        pass
+
+    @parameters.setter
+    def parameters(self, parameters):
+        """Set parameters of the Model."""
+        pass
+
+
+class AutoPickableModel(BaseModel, metaclass=abc.ABCMeta):
+    """
+    Abstract class for models with automatic pickling.
+
+    This class follows the BaseModel API and handles pickling automatically.
+    It provides both self.init_spec and self.parameters for pickling, which
+    are the arguments and weights of the Model. When unpickled, a new model
+    will be reconstructed with self.init_spec as arguments and load the weights
+    from self.parameters.
+
+    The target weights should be assigned to self._default_parameters during
+    pickling, so that the newly created model can check if target weights exist
+    or not.
+    """
+
+    def __init__(self):
+        self._name = type(self).__name__  # name default to class
+        self._inputs = None
+        self._outputs = None
+        self._default_parameters = None
+
+    @property
+    def init_spec(self):
```

Comment:
I think we should replace this function with something similar to `quick_init()`. This will prevent us from accidentally losing user data in the future, which is something we promise not to do.

Reply:
i would prefer something less magical like your proposal here https://github.com/rlworkgroup/proposals/pull/3

Reply:
I would also be perfectly happy with that. The important thing is that we don't set ourselves up to shoot feet (our own or otherwise).

Reply:
i'm going to resolve all but one comment which addressing pickling strategy